src/main/java/teammates/ui/controller/InstructorStudentListPageAction.java

```diff
@@ -21,32 +24,45 @@ public ActionResult execute() throws EntityDoesNotExistException {
-
-        data = new InstructorStudentListPageData(account);
-        data.instructors = new HashMap<String, InstructorAttributes>();
-        data.numStudents = new HashMap<String, String>();
-        data.courses = logic.getCoursesForInstructor(account.googleId);
-        for (CourseAttributes course : data.courses) {
+        Map<String, InstructorAttributes> instructors = new HashMap<String, InstructorAttributes>();
+        Map<String, String> numStudents = new HashMap<String, String>();
+        List<CourseAttributes> courses = logic.getCoursesForInstructor(account.googleId);
+        for (CourseAttributes course : courses) {
             InstructorAttributes instructor = logic.getInstructorForGoogleId(course.id, account.googleId);
-            data.instructors.put(course.id, instructor);
+            instructors.put(course.id, instructor);
             int numStudentsInCourse = logic.getStudentsForCourse(course.id).size();
-            data.numStudents.put(course.id, String.valueOf(numStudentsInCourse));
+            numStudents.put(course.id, String.valueOf(numStudentsInCourse));
         }
-
-        Collections.sort(data.courses, new Comparator<CourseAttributes>() {
+        Collections.sort(courses, new Comparator<CourseAttributes>() {
             @Override
             public int compare(CourseAttributes c1, CourseAttributes c2) {
                 return c1.createdAt.compareTo(c2.createdAt);
             }
         });
-        data.searchKey = searchKey;
-        data.displayArchive = displayArchive;
 
-        if (data.courses.size() == 0) {
+        if (courses.size() == 0) {
             statusToUser.add(Const.StatusMessages.INSTRUCTOR_NO_COURSE_AND_STUDENTS);
         }
 
-        statusToAdmin = "instructorStudentList Page Load<br>" + "Total Courses: " + data.courses.size();
+        statusToAdmin = "instructorStudentList Page Load<br>" + "Total Courses: " + courses.size();
+        
+        data = new InstructorStudentListPageData(account);
+        List<InstructorStudentListPageCourseData> coursesToDisplay = new ArrayList<InstructorStudentListPageCourseData>();
+        for (CourseAttributes course: courses) {
+            InstructorAttributes instructor = instructors.get(course.id);
+            boolean isInstructorAllowedToModify = instructor.isAllowedForPrivilege(
+                                            Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_STUDENT);
+            boolean isCourseArchived = data.isCourseArchived(course, instructor);
+            boolean isCourseDisplayed = displayArchive || !isCourseArchived;
+            if (isCourseDisplayed) {
+                coursesToDisplay.add(new InstructorStudentListPageCourseData(course, isCourseArchived,
+                                                                             isInstructorAllowedToModify,
+                                                                        data.getInstructorCourseEnrollLink(course.id)));
```

Comment:
As this segment does not involve `Logic`, perhaps it should be placed in `*PageData` instead?